### PR TITLE
Readability improvements for two tutorial notebooks.

### PR DESCRIPTION
**Pull request by delpart**
### Summary
Code updates for the two tutorials `02_neural_network_classifier_and_regressor.ipynb` and `05_torch_connector.ipynb` to increase readability.

### Details and comments
Replaced `for-loops` with appropriate NumPy or Tensor operations, where appropriate, which should make the code more concise.

### CODE REVIEW

**Comment by woodsp-ibm** [comment]
The output of cell [6], in the current version of the Torch Connector NOtebook shows an accuracy of 0.8 with 3 blue and 1 green dot ringed with red - how it comes out in your changed version is rather different

image

**Comment by delpart** [comment]
My bad, sorry about that. That went under my radar. Fixed it with 28ce229.

image

File: docs/tutorials/05_torch_connector.ipynb [hunk 1]
```diff
@@ -114,11 +114,8 @@
     "y_ = Tensor(y).reshape(len(y), 1)\n",
     "\n",
     "# Plot dataset\n",
-    "for x, y_target in zip(X, y):\n",
-    "    if y_target == 1:\n",
-    "        plt.plot(x[0], x[1], 'bo')\n",
-    "    else:\n",
-    "        plt.plot(x[0], x[1], 'go')\n",
+    "plt.plot(X[y == 1, 0], X[y == 1, 1], 'bo')\n",
+    "plt.plot(X[y != 1, 0], X[y != 1, 1], 'go')\n",
```
Comment (adekusar-drl) [inline docs/tutorials/05_torch_connector.ipynb:118]: I understand what you want to do here, but usually tutorials intend to be simple, while some true pythonic stuff may puzzle a reader, e.g. `X[y == 1, 0], X[y == 1, 1]`. My suggestion is either get back to the original code or add a comment for each line explaining what is written there and how it works.

File: docs/tutorials/05_torch_connector.ipynb [hunk 2]
```diff
@@ -311,22 +310,15 @@
    ],
    "source": [
     "# Evaluate model and compute accuracy\n",
-    "y_predict = []\n",
-    "for x, y_target in zip(X, y):\n",
-    "    output = model1(Tensor(x))\n",
-    "    y_predict += [np.sign(output.detach().numpy())[0]]\n",
-    "\n",
+    "y_predict = np.sign(model1(X_).detach().numpy()).squeeze()\n",
+    "    \n",
```
Comment (adekusar-drl) [inline docs/tutorials/05_torch_connector.ipynb:314]: In general, the same comment, the line `np.sign(model1(X_).detach().numpy()).squeeze()` has a lot operations and may take some time to understand. I'm not saying that the original code is way better, but since we change it is nice to have make it really nice. Would be valuable to add a comment like "we evaluate model with a batch of data... " or something like that and explain why `squeeze` is required.

File: docs/tutorials/05_torch_connector.ipynb [hunk 3]
```diff
@@ -311,22 +310,15 @@
...
-    "for x, y_target, y_p in zip(X, y, y_predict):\n",
-    "    if y_target == 1:\n",
-    "        plt.plot(x[0], x[1], 'bo')\n",
-    "    else:\n",
-    "        plt.plot(x[0], x[1], 'go')\n",
-    "    if y_target != y_p:\n",
-    "        plt.scatter(x[0], x[1], s=200, facecolors='none', edgecolors='r', linewidths=2)\n",
+    "plt.plot(X[y == 1, 0], X[y == 1, 1], 'bo')\n",
+    "plt.plot(X[y != 1, 0], X[y != 1, 1], 'go')\n",
+    "plt.scatter(X[y != y_predict, 0], X[y != y_predict, 1], s=200, facecolors='none', edgecolors='r', linewidths=2)\n",
```
Comment (adekusar-drl) [inline docs/tutorials/05_torch_connector.ipynb:321]: Same as above.

File: docs/tutorials/05_torch_connector.ipynb [hunk 4]
```diff
@@ -485,22 +473,15 @@
...
    "source": [
     "# Evaluate model and compute accuracy\n",
-    "y_predict = []\n",
-    "for x in X:\n",
-    "    output = model2(Tensor(x))\n",
-    "    y_predict += [np.argmax(output.detach().numpy())]\n",
-    "\n",
-    "print('Accuracy:', sum(y_predict == y01)/len(y01))\n",
+    "y_predict = np.argmax(model2(X_).detach().numpy(), axis=1)\n",
+    "    \n",
```
Comment (adekusar-drl) [inline docs/tutorials/05_torch_connector.ipynb:477]: Same as above

File: docs/tutorials/05_torch_connector.ipynb [hunk 5]
```diff
@@ -485,22 +473,15 @@
...
-    "for x, y_target, y_ in zip(X, y01, y_predict):\n",
-    "    if y_target == 1:\n",
-    "        plt.plot(x[0], x[1], 'bo')\n",
-    "    else:\n",
-    "        plt.plot(x[0], x[1], 'go')\n",
-    "    if y_target != y_:\n",
-    "        plt.scatter(x[0], x[1], s=200, facecolors='none', edgecolors='r', linewidths=2)\n",
+    "plt.plot(X[y01 == 1, 0], X[y01 == 1, 1], 'bo')\n",
+    "plt.plot(X[y01 != 1, 0], X[y01 != 1, 1], 'go')\n",
+    "plt.scatter(X[y01 != y_predict, 0], X[y01 != y_predict, 1], s=200, facecolors='none', edgecolors='r', linewidths=2)\n",
```
Comment (adekusar-drl) [inline docs/tutorials/05_torch_connector.ipynb:484]: Same as above

File: docs/tutorials/05_torch_connector.ipynb [hunk 6]
```diff
@@ -717,10 +692,7 @@
     "plt.plot(X, y, 'bo')\n",
     "\n",
     "# Plot fitted line\n",
-    "y_ = []\n",
-    "for x in np.linspace(lb, ub):\n",
-    "    output = model3(Tensor([x]))\n",
-    "    y_ += [output.detach().numpy()[0]]\n",
+    "y_ = model3(Tensor(np.linspace(lb, ub)).unsqueeze(-1)).detach().numpy().squeeze()\n",
```
Comment (adekusar-drl) [inline docs/tutorials/05_torch_connector.ipynb:695]: Over-complicated line for the tutorial, requires explanations. I suggest at least split it into two, e.g.
```python
input_data = Tensor(np.linspace(lb, ub)).unsqueeze(-1)
y_ = model3(input_data).detach().numpy().squeeze()
```
and add comments.

File: docs/tutorials/05_torch_connector.ipynb [hunk 7]
```diff
@@ -833,9 +805,9 @@
     "data_iter = iter(train_loader)\n",
     "fig, axes = plt.subplots(nrows=1, ncols=n_samples_show, figsize=(10, 3))\n",
     "\n",
-    "while n_samples_show > 0:\n",
-    "    images, targets = data_iter.__next__()\n",
-    "\n",
+    "for images, targets in data_iter:\n",
+    "    if n_samples_show <= 0:\n",
+    "        break\n",
```
Comment (adekusar-drl) [inline docs/tutorials/05_torch_connector.ipynb:810]: Would it be possible to change the `if` statement to `if count == n_samples_show:` just to consistency with another cell?

File: docs/tutorials/02_neural_network_classifier_and_regressor.ipynb [hunk 8]
```diff
@@ -95,14 +95,11 @@
...
-    "for x, y_target in zip(X, y):\n",
-    "    if y_target == 1:\n",
-    "        plt.plot(x[0], x[1], 'bo')\n",
-    "    else:\n",
-    "        plt.plot(x[0], x[1], 'go')\n",
+    "y_one_hot[y01 == 1] = [0, 1]\n",
+    "y_one_hot[y01 == 0] = [1, 0]\n",
+    "\n",
+    "plt.plot(X[y == 1, 0], X[y == 1, 1], 'bo')\n",
+    "plt.plot(X[y != 1, 0], X[y != 1, 1], 'go')\n",
```
Comment (adekusar-drl) [inline docs/tutorials/02_neural_network_classifier_and_regressor.ipynb:102]: Same issue as in the torch connector tutorial. I'd say too pythonic and may require additional explanation to a user.

File: docs/tutorials/02_neural_network_classifier_and_regressor.ipynb [hunk 9]
```diff
@@ -209,17 +204,13 @@
...
-    "for x, y_target, y_p in zip(X, y, y_predict):\n",
-    "    if y_target == 1:\n",
-    "        plt.plot(x[0], x[1], 'bo')\n",
-    "    else:\n",
-    "        plt.plot(x[0], x[1], 'go')\n",
-    "    if y_target != y_p:\n",
-    "        plt.scatter(x[0], x[1], s=200, facecolors='none', edgecolors='r', linewidths=2)\n",
+    "plt.plot(X[y == 1, 0], X[y == 1, 1], 'bo')\n",
+    "plt.plot(X[y != 1, 0], X[y != 1, 1], 'go')\n",
+    "plt.scatter(X[y != y_predict, 0], X[y != y_predict, 1], s=200, facecolors='none', edgecolors='r', linewidths=2)\n",
```
Comment (adekusar-drl) [inline docs/tutorials/02_neural_network_classifier_and_regressor.ipynb:213]: Same as above

File: docs/tutorials/02_neural_network_classifier_and_regressor.ipynb [hunk 10]
```diff
@@ -353,17 +354,13 @@
...
-    "for x, y_target, y_p in zip(X, y01, y_predict):\n",
-    "    if y_target == 1:\n",
-    "        plt.plot(x[0], x[1], 'bo')\n",
-    "    else:\n",
-    "        plt.plot(x[0], x[1], 'go')\n",
-    "    if y_target != y_p:\n",
-    "        plt.scatter(x[0], x[1], s=200, facecolors='none', edgecolors='r', linewidths=2)\n",
+    "plt.plot(X[y01 == 1, 0], X[y01 == 1, 1], 'bo')\n",
+    "plt.plot(X[y01 != 1, 0], X[y01 != 1, 1], 'go')\n",
+    "plt.scatter(X[y01 != y_predict, 0], X[y01 != y_predict, 1], s=200, facecolors='none', edgecolors='r', linewidths=2)\n",
```
Comment (adekusar-drl) [inline docs/tutorials/02_neural_network_classifier_and_regressor.ipynb:363]: Same as above

File: docs/tutorials/02_neural_network_classifier_and_regressor.ipynb [hunk 11]
```diff
@@ -443,17 +440,13 @@
...
-    "for x, y_target, y_p in zip(X, y_one_hot, y_predict):\n",
-    "    if y_target[0] == 1:\n",
-    "        plt.plot(x[0], x[1], 'bo')\n",
-    "    else:\n",
-    "        plt.plot(x[0], x[1], 'go')\n",
-    "    if not np.all(y_target == y_p):\n",
-    "        plt.scatter(x[0], x[1], s=200, facecolors='none', edgecolors='r', linewidths=2)\n",
+    "plt.plot(X[y01 == 1, 0], X[y01 == 1, 1], 'bo')\n",
+    "plt.plot(X[y01 != 1, 0], X[y01 != 1, 1], 'go')\n",
+    "plt.scatter(X[y01 != y_predict, 0], X[y01 != y_predict, 1], s=200, facecolors='none', edgecolors='r', linewidths=2)\n",
```
Comment (adekusar-drl) [inline docs/tutorials/02_neural_network_classifier_and_regressor.ipynb:449]: Same as above

**Review by adekusar-drl — CHANGES_REQUESTED** [review]
I really like what you've done, especially adding a batch mode in the torch connector tutorial. But some changes may be uneasy to read for newcomers to python. A few comments could really help readers to understand what's going on. Thanks a lot!

File: docs/tutorials/05_torch_connector.ipynb [hunk 12]
```diff
@@ -717,10 +783,8 @@
     "plt.plot(X, y, 'bo')\n",
     "\n",
     "# Plot fitted line\n",
-    "y_ = []\n",
-    "for x in np.linspace(lb, ub):\n",
-    "    output = model3(Tensor([x]))\n",
-    "    y_ += [output.detach().numpy()[0]]\n",
+    "input_tensor = Tensor(np.linspace(lb, ub)).unsqueeze(-1)\n",
+    "y_ = model3(input_tensor).cpu().detach().numpy().squeeze()\n",
```
Comment (adekusar-drl) [inline docs/tutorials/05_torch_connector.ipynb:787]:
```suggestion
    "# generate a random tensor and restore an additional batch dimension by unsqueezing it.",
    "input_tensor = Tensor(np.linspace(lb, ub)).unsqueeze(-1)\n",
    "# remove batch dimension and make the output array flat",
    "y_ = model3(input_tensor).cpu().detach().numpy().squeeze()\n",
```

**Review by adekusar-drl — COMMENTED** [review]
Thanks a lot for the improvements, please review the suggestions, I hope, they make the notebooks better.

**Comment by adekusar-drl** [comment]
I close this PR as it is stalled and I don't see much benefits.
